support/reftest/reftest

```diff
@@ -0,0 +1,296 @@
+    parser.add_argument(
+        "--cert",
+        default="~/certs/rcm-debug/rcm-debug.crt",
+        type=str,
+        help="Client certificate file and password",
+    )
+    parser.add_argument(
+        "--key",
+        default="~/certs/rcm-debug/rcm-debug.key",
+        type=str,
+        help="Private key file name",
+    )
+    parser.add_argument(
+        "--cacert",
+        default="/etc/rhsm/ca/redhat-uep.pem",
+        type=str,
+        help="CA certificate to verify peer against",
+    )
+
+
+def parse_args():
+    root_parser = argparse.ArgumentParser(description=__doc__)
+    subparsers = root_parser.add_subparsers(
+        dest="command", help="reference test data operations"
+    )
+    parser_prepare = subparsers.add_parser(
+        "prepare",
+        help="""
+        prepare reference test data for integration test
+         e.g. $./reftest prepare --bucket dichen-test --table dichen-test
+        """,
+    )
+    parser_prepare.add_argument(
+        "--release-date",
+        default=None,
+        help="Date on which the content will be made available.",
+    )
+    parse_aws_info(parser_prepare)
+    parse_cdn_certs(parser_prepare)
+
+    parser_clean = subparsers.add_parser(
+        "clean",
+        help="""
+        clean reference test data in s3 and dynamodb
+        e.g. $./reftest clean --bucket dichen-test --table dichen-test
+        """,
+    )
+    parse_aws_info(parser_clean)
+
+    return root_parser.parse_args()
+
+
+class RefTestConfig:
+    def __init__(self, prod_cdn_url, test_data):
+        self.prod_cdn_url = prod_cdn_url
+        self.test_data = test_data
+
+
+def load_config():
+    with open("data.yml") as f:
+        config = yaml.load(f, yaml.SafeLoader)
+    return RefTestConfig(config["prod-cdn-url"], config["test_data"],)
+
+
+def download_to_local(url, key_path, cert_path, cacert_path):
+    with requests.get(
+        url,
+        cert=(cert_path, key_path),
+        verify=cacert_path,
+        stream=True,
+        timeout=(30, 30),
+    ) as r:
+        r.raise_for_status()
+        total_size = int(r.headers.get("content-length", 0))
+        t = tqdm(desc=url, total=total_size, unit="iB", unit_scale=True)
+
+        # the file is deleted as soon as it is closed.
+        temp_file = tempfile.NamedTemporaryFile(delete=True)
+        for chunk in r.iter_content(chunk_size=8192):
+            if chunk:  # filter out keep-alive new chunks
+                t.update(len(chunk))
+                temp_file.write(chunk)
+        t.close()
+        return temp_file
+
+
+def prepare(db, s3, config, opt):
+    for item in config.test_data:
+        url = config.prod_cdn_url + item["path"]
+        # download test data to local with a name "NamedTemporaryFile"
+        temp_file = download_to_local(url, opt.key, opt.cert, opt.cacert)
+        # push test data to s3 and dynamodb
+        s3.upload_from_localfile(
+            temp_file.name, item["sha256"], item["content-type"]
+        )
```

Comment:
Related to my comment on data.yaml, we may want to calculate the temp_file's checksum to use here instead of betting that it'll be consistent.

Reply:
+1

This is pretty important to detect if reftest.yml has become outdated.

Reply:
Consensus was that we should use temp_file's checksum.

Reply:
Just to clarify as there still seems to be a bit of confusion here, what I think this needs to do is:

- calculate the temp file's checksum
- verify that the file's checksum matches item["sha256"] - if not, it's a fatal error

Obviously then when you actually do the insert, it doesn't matter whether you use item["sha256"] or the temp file's checksum because you made sure that they are the same value.